File: Recursion Maze/Backtracking/Q3. The k-th Lexicographical String of All Happy Strings of Length n/Q3_Thek-thLexicographicalStringofAllHappyStringsofLengthn.py
class Solution(object):
    def getHappyString(self, n, k):
        """
        :type n: int      # Length of happy string
        :type k: int      # We need the k-th happy string
        :rtype: str       # Return type is string
        """
        
        # DFS function to build happy strings
        def dfs(path=''):
            
            # If we already found k strings, stop recursion (optimization)
            if len(res) == k:
                return
            
            # If current string length becomes n,
            # it means we formed a valid happy string
            if len(path) == n:
                return res.append(path)   # Store it in result list
            
            # Try all possible characters
            for x in 'abc':
                
                # Add character only if:
                # 1) path is empty OR
                # 2) current character is not same as previous character
                # (because happy string cannot have consecutive same letters)
                if not path or x != path[-1]:
                    dfs(path + x)   # Recursively build next character
        
        res = []   # List to store valid happy strings
        
        dfs()      # Start DFS with empty string
        
        # If we found at least k happy strings,
        # return the k-th one (last added)
        # Otherwise return empty string
        return res[-1] if len(res) == k else ''

File: Recursion Maze/Backtracking/Q3. The k-th Lexicographical String of All Happy Strings of Length n/Q3_Thek-thLexicographicalStringofAllHappyStringsofLengthn.java
class Solution {
    
    int n2;   // Stores original length n (used for calculating remaining combinations)
    
    public String getHappyString(int n, int k) {
        n2 = n;   // Save original n
        return dfs(new StringBuilder(), n, k);  // Start DFS with empty prefix
    }
    
    public String dfs(StringBuilder prefix, int n, int k){
        
        // Base case: if no more characters to add
        if (n == 0)
            return prefix.toString();   // Return completed happy string
        
        // Try characters from 'a' to 'c'
        for (char c = 'a'; c <= 'c'; c++) {
            
            // Skip if same as previous character (to maintain happy string condition)
            if (prefix.length() > 0 && c == prefix.charAt(prefix.length() - 1))
                continue;
            
            // Count how many valid strings can be formed
            // if we choose this character now
            int cnt = (int) Math.pow(2, n2 - prefix.length() - 1);
            
            // If the k-th string lies in this block
            if (cnt >= k)
                return dfs(prefix.append(c), n - 1, k);  // Choose this char and go deeper
            
            else
                k -= cnt;  // Skip this block and reduce k
        }
        
        return "";  // If no valid string found, return empty
    }
}